rfcs/20180824-tf-print-v2.md

```diff
@@ -0,0 +1,549 @@
+1.  Because printing & formatting follows the current device context, printing
+    may happen on a different machine than users intend, and then they have a
+    hard time finding the printed data. Or, it may unintentionally transfer a
+    large tensor over the network before formatting it.
+    *   Developers may be more inclined to have printing/logging happening in
+        their client as opposed to on various machines in a distributed system.
+1.  To format & print on different devices, users must explicitly call both
+    `tf.strings.format` and `tf.print`
+1.  Someone using normal python print in eager mode python notebooks or repl
+    might get used to it working, and not realize that there's a separate
+    `tf.print`. They may then try using it in a Graph Function and have their
+    printing not work. In colab they might switch to tf.print then suddenly be
+    confused why nothing is printing.
+    *   It may make sense for eager graph functions to automatically hook the
+        python print method to replace it with calls to `tf.Print`. This would
+        still be problematic in python notebooks though (for above reasons).
+1.  Various user code could still end up calling EagerTensor `__str__`
+    unintentionally, which calls .numpy() on the tensor then prints it. This
+    will copy the full tensor over then format it w/ numpy, which will have
+    similar but slightly mismatched formatting from tf.print.
+    *   For consistency, maybe make `__str__` and `__repr__` in EagerTensors use
+        the tf.strings.format op on whatever device the tensor is on? This could
+        avoid transferring giant tensors over the network unintentionally.
+1.  The `tf.print` functionality to deal w/ arbitrary nested structures will be
+    python-specific. Other language bindings will have to manually chain format
+    & print.
+
+## Extra Details
+
+### Supported Input Types
+
+Any Python object can be passed as input to `tf.print`, and it will be printed.
+
+Tensors will automatically be extracted from python lists, dicts, and other data
+structures supported by the TensorFlow `nest` utilities. These extracted tensors
+will print with their summarization format whenever `tf.print` executes.
+TensorFlow variables also work correctly inside of `tf.print` (even when nested
+inside data structures).
+
+However, there are the following caveats!
+
+*   In session mode and inside graph functions (and possibly in eager loops as
+    well), the printed value of non-tensor python objects will always match what
+    it was the very first time that a given `tf.print` is executed. It will not
+    update when that specific `tf.print` is called again.
+*   This also means that the structure of printed lists/dicts is also only
+    captured the first time a `tf.print` is executed. If tensors are added or
+    removed to a list printed in `tf.print` multiple times, those changes to the
+    list will not be captured when printing again!
```

Comment:
This seems like it will be a confusing gotcha for users. Is there some way to indicate that the value is stale?

Reply:
I just ran this by @akshayka, and this actually shouldn't be an issue.
Non-eager tf control flow doesn't support modifying python objects anyway, and the tfe.defun graph functions should automatically detect that the cached value is stale and will re-generate the ops.

I'll remove this caveat from the doc.